src/test/java/spoon/test/template/PatternTest.java

```diff
@@ -0,0 +1,1296 @@
+			assertEquals(false, match.getParameters().getValue("option"));
+			assertEquals(true, match.getParameters().getValue("option2"));
+			assertEquals("i", match.getParameters().getValue("value").toString());
+		}
+		{
+			Match match = matches.get(1);
+			assertEquals(Arrays.asList("java.lang.System.out.println(\"a\")"), listToListOfStrings(match.getMatchingElements()));
+			assertEquals(true, match.getParameters().getValue("option"));
+			assertEquals(false, match.getParameters().getValue("option2"));
+			assertEquals("\"a\"", match.getParameters().getValue("value").toString());
+		}
+		{
+			Match match = matches.get(2);
+			assertEquals(Arrays.asList("java.lang.System.out.println(\"Xxxx\")"), listToListOfStrings(match.getMatchingElements()));
+			assertEquals(true, match.getParameters().getValue("option"));
+			assertEquals(false, match.getParameters().getValue("option2"));
+			assertEquals("\"Xxxx\"", match.getParameters().getValue("value").toString());
+		}
+		{
+			Match match = matches.get(3);
+			assertEquals(Arrays.asList("java.lang.System.out.println(((java.lang.String) (null)))"), listToListOfStrings(match.getMatchingElements()));
+			assertEquals(true, match.getParameters().getValue("option"));
+			assertEquals(false, match.getParameters().getValue("option2"));
+			assertEquals("((java.lang.String) (null))", match.getParameters().getValue("value").toString());
+		}
+		{
+			Match match = matches.get(4);
+			assertEquals(Arrays.asList("java.lang.System.out.println(2018)"), listToListOfStrings(match.getMatchingElements()));
+			assertEquals(false, match.getParameters().getValue("option"));
+			assertEquals(true, match.getParameters().getValue("option2"));
+			assertEquals("2018", match.getParameters().getValue("value").toString());
+		}
+		{
+			Match match = matches.get(5);
+			assertEquals(Arrays.asList("java.lang.System.out.println(java.lang.Long.class.toString())"), listToListOfStrings(match.getMatchingElements()));
+			assertEquals(true, match.getParameters().getValue("option"));
+			assertEquals(false, match.getParameters().getValue("option2"));
+			assertEquals("java.lang.Long.class.toString()", match.getParameters().getValue("value").toString());
+		}
+		{
+			Match match = matches.get(6);
+			assertEquals(Arrays.asList("java.lang.System.out.println(3.14)"), listToListOfStrings(match.getMatchingElements()));
+			assertEquals(false, match.getParameters().getValue("option"));
+			assertEquals(false, match.getParameters().getValue("option2"));
+			assertEquals("3.14", match.getParameters().getValue("value").toString());
+		}
+	}
+	@Test
+	public void testGenerateMultiValues() throws Exception {
+		// contract: the pattern parameter (in this case 'statements')
+		//can have type List and can be replaced by list of elements
+		//(in this case by list of statements)
+
+		// here, in particular, we test method "substituteList"

```

Comment:
use substituteList when you expect 0, 1 or more generated root elements
use substituteSingle when you expect exactly one generated root element

Note: internally both cases are handled same, here is the code from `spoon.pattern.Generator'
```java
	/**
	 * Generates one target depending on kind of this {@link RootNode}, expected `expectedType` and input `parameters`
	 * @param node to be generated node
	 * @param parameters {@link ParameterValueProvider}
	 * @param expectedType defines {@link Class} of returned value
	 *
	 * @return a generate value or null
	 */
	default <T> T generateTarget(RootNode node, ParameterValueProvider parameters, Class<T> expectedType) {
		ResultHolder.Single<T> result = new ResultHolder.Single<>(expectedType);
		generateTargets(node, result, parameters);
		return result.getResult();
	}

	/**
	 * Generates zero, one or more targets depending on kind of this {@link RootNode}, expected `expectedType` and input `parameters`
	 * @param node to be generated node
	 * @param parameters {@link ParameterValueProvider}
	 * @param expectedType defines {@link Class} of returned value
	 *
	 * @return a {@link List} of generated targets
	 */
	default <T> List<T> generateTargets(RootNode node, ParameterValueProvider parameters, Class<T> expectedType) {
		ResultHolder.Multiple<T> result = new ResultHolder.Multiple<>(expectedType);
		generateTargets(node, result, parameters);
		return result.getResult();
	}
```

Reply:
Here? I have a concern with the API: we have 4 public methods `substitute*`. It is hard for client to quickly grasp the differences. Is it possible to have a single `substituteMethod?`

Reply:
We can easily get rid of the methods with `Map` parameters. Then `substitueList` is more generic so it can survive.

But I would prefer to keep all 4. Client can use the one which fits best to its needs. I have experience with APIs of other SW and I always approaciate when I as client can choose the best fitting method.